[extractor/prankcast] Add extractor

columndeeply opened the pull request:
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

<!--

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible

-->

Added a new extractor for prankcast.com

Fixes #


<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--

# PLEASE FOLLOW THE GUIDE BELOW

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [X] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

pukkandan commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,55 @@
+        webpage = self._download_webpage(url, video_id)
+
+        # Extract the JSON
+        json = self._html_search_regex(r'<script id=\"__NEXT_DATA__\"[^>]*>(.*)</script>', webpage, 'json_info')
```
Use _search_nextjs_data

columndeeply replied:
Done

columndeeply commented:
Quick question since it's still in review: is there any field were I could add extra "general" info? For example I'd like to add the list of cohosts but that wouldn't fit in the "uploader" field since they aren't the host of the show. Is there a "notes" field for this type of think or can I just return a new field "cohosts" with that info?

pukkandan commented:
> * [x]  At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-08-27T10%3A27%3A56Z#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-08-27T10%3A27%3A56Z#yt-dlp-coding-conventions)

Re-read https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#mandatory-and-optional-metafields

columndeeply commented:
Done, now it returns None on the optional fields if there's no data.

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,79 @@
+        webpage = self._download_webpage(url, video_id)
+
+        # Extract the JSON
+        json_info = self._search_nextjs_data(webpage, video_id)
```
```suggestion
        json_info = self._search_nextjs_data(webpage, video_id)['props']['pageProps']['ssr_data_showreel']
```

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,79 @@
+        url = broadcast_url + recording_hash + ".mp3"
+
+        # Get dates
+        start_date = None
+        if 'start_date' in json_info['props']['pageProps']['ssr_data_showreel']:
```
This block of code can rewritten something like:
``` var = json_info.get('key')```

HobbyistDev replied:
This line (line 34-35) shouldn't fatal, right ?. Use `traverse_obj` from `utils`

columndeeply replied:
Is using ".get(key)" like I did in the last commit okay? "traverse_obj" seems redundant since I'm already getting the "['props']['pageProps']['ssr_data_showreel']" path when parsing the "json_info" (using the changes you proposed above). If "traverse_obj" is still required let me know and I'll change it, the contribution guide says both are OK but I'm not sure when to use one or the other...

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,79 @@
+        if 'start_date' in json_info['props']['pageProps']['ssr_data_showreel']:
+            start_date = json_info['props']['pageProps']['ssr_data_showreel']['start_date'].replace('Z', '')
+
+        end_date = None
+        if 'end_date' in json_info['props']['pageProps']['ssr_data_showreel']:
```
Same as above

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,79 @@
+            end_date = json_info['props']['pageProps']['ssr_data_showreel']['end_date'].replace('Z', '')
+
+        # Get broadcast date
+        upload_date = None
+        if start_date is not None:
```
Same as above

columndeeply replied:
Can't use ".get(key)" here, start_date is just a string. It's not part of the "json_info"

HobbyistDev replied:
My bad

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,79 @@
+        broadcast_title = json_info['props']['pageProps']['ssr_data_showreel']['broadcast_title']
+
+        # Get author (AKA show host)
+        uploader = None
+        if 'user_name' in json_info['props']['pageProps']['ssr_data_showreel']:
```
Same as above

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,79 @@
+            guests = []
+
+        if 'guests_json' in json_info['props']['pageProps']['ssr_data_showreel']:
+            for guest in json.loads(json_info['props']['pageProps']['ssr_data_showreel']['guests_json']):
```
I think `_parse_json` can be used here instead of `json.loads`. But i think `json_info['props']['pageProps']['ssr_data_showreel']['guests_json']` already in json, no ?

columndeeply replied:
Changed it to _parse_json but I do have to parse it again. Otherwise I get a string and can't use the loop on it.

On the original json it looks like this:
`"guests_json":"[{\"id\": 2, \"date\": \"2022-08-08T23:56:19.000Z\", \"name\": \"dragonmere\", \"open\": 0, \"type\": \"cohost\", \"timestamp\": \"\", \"permissions\": \"0\"}, {\"id\": 14, \"date\": \"2022-08-08T23:56:25.000Z\", \"name\": \"splink\", \"open\": 0, \"type\": \"guest\", \"timestamp\": \"\", \"permissions\": \"0\"}, {\"id\": 19, \"date\": \"2022-08-08T23:56:35.000Z\", \"name\": \"wastedmemory\", \"open\": 0, \"type\": \"guest\", \"timestamp\": \"\", \"permissions\": \"0\"}, {\"id\": 585, \"date\": \"2022-08-09T00:57:36.000Z\", \"name\": \"bread336\", \"open\": 0, \"type\": \"guest\", \"timestamp\": \"\", \"permissions\": \"0\"}, {\"id\": 373, \"date\": \"2022-08-09T01:27:34.000Z\", \"name\": \"criefaux\", \"open\": 0, \"type\": \"guest\", \"timestamp\": \"\", \"permissions\": \"0\"}, {\"id\": 277, \"date\": \"2022-08-09T01:27:42.000Z\", \"name\": \"kbbles\", \"open\": 0, \"type\": \"guest\", \"timestamp\": \"\", \"permissions\": \"0\"}]",`

HobbyistDev replied:
Could you try `_parse_json` with argument `transform_source=js_to_json`? Be sure to import `js_to_json` first

columndeeply replied:
Sorry, I got it wrong on the last comment. Using _parse_json works great (without needing the js_to_json), not sure why it was giving me a string before...

HobbyistDev reviewed (changes requested):
Please re-read the contribution guide

columndeeply commented:
I've applied all the changes and added a fallback to the title, I don't see anything else from the contributions guidelines that's missing...

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,77 @@
+        url = broadcast_url + recording_hash + ".mp3"
+
+        # Get broadcast title
+        broadcast_title = json_info.get('broadcast_title') or self._og_search_title(webpage)
```
inline in return

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,77 @@
+        broadcast_title = json_info.get('broadcast_title') or self._og_search_title(webpage)
+
+        # Get dates
+        start_date = json_info.get('start_date')
+        if 'start_date' != '':
+            start_date = start_date.replace('Z', '')
+
+        end_date = json_info.get('end_date')
+        if 'end_date' != '':
+            end_date = end_date.replace('Z', '')
```
Please correct me if i'm wrong, i think `start_date` and `end_date` can parsed with `parse_iso8601` from `utils`

columndeeply replied:
Yeah, you're right. I'll change these to use parse_iso8601

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,77 @@
+
+
+class PrankCastIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?prankcast\.com/.*/showreel/(?P<id>[0-9]+)'
```
use `\d`

HobbyistDev replied:
i think the string after `id` can be used to be `display_id`

HobbyistDev commented:
You can get `tags` and `categories` from `broadcast_tags` and `broadcast_category`

columndeeply commented:
Right, I believe it's always the same category but I'll add it just in case I'm wrong.

I'll push a commit when I get back from work with the tags/categories added + changes for the rest of the comments.

columndeeply commented:
Done, here's the latest changes:
- Added `display_id` using the string after the `video_id`
- Add `tags`, `categories`, `description`, `channel_id`
- Moved the date stuff to use `parse_iso8601`
- Returns `timestamp` instead of `upload_date` since it'll be using that field anyways
- Move the `cast` to a real list instead of a string

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -17,7 +16,9 @@ class PrankCastIE(InfoExtractor):
 
     def _real_extract(self, url):
-        video_id = self._match_id(url)
+        match = self._match_valid_url(url).groupdict()
+        video_id, display_id = match['id'], match['display_id']
```
You can change this line like this:'
```python
video_id, display_id = self._match_valid_url(url).group('id', 'display_id')
```

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -29,49 +30,48 @@ def _real_extract(self, url):
-        if start_date is not None:
-            upload_date = start_date.split('T')[0].replace('-', '')
+        # Get broadcast info
+        broadcast_title = json_info.get('broadcast_title')
+        broadcast_description = json_info.get('broadcast_description')
+        broadcast_category = json_info.get('broadcast_category')
+        broadcast_tags = self._parse_json(json_info.get('broadcast_tags'), video_id)
```
inline in return

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,71 @@
+
+        # Get the broadcast URL and the recording hash.
+        # The full URL is {broadcast_url}/{recording_hash}.mp3
+        broadcast_url = json_info.get('broadcast_url')
+        recording_hash = json_info.get('recording_hash')
+        url = broadcast_url + recording_hash + ".mp3"
```
I think this should fatal. fstring is more preferred than `+` sign. I think you can rewrite it like this :
```suggestion
        url = f'{json_info["broadcast_url"]}{json_info["recording_hash"]}.mp3'
```
Even better, you can inline this at return

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,71 @@
+        else:
+            guests = []
+
+        guests_json = json_info.get('guests_json')
+        if guests_json is not None:
+            guests_json = self._parse_json(guests_json, video_id)
+            for guest in guests_json:
+                guest_name = guest.get('name')
+                if guest_name is not None:
+                    guests.append(guest_name)
```
I think list comperhension can be used here

HobbyistDev commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,71 @@
+        uploader = json_info.get('user_name')
+
+        # Get the co-hosts/guests
+        if uploader != '':
+            guests = [uploader]
+        else:
+            guests = []
```
This can be rewrite as:
```suggestion
            guests = [] if not uploader else [uploader]
```
probably, you can use list comperhension for guest value

columndeeply commented:
Done, not sure if there's a cleaner way to do the "cast" part, let me know if it could be improved. I've moved the URL part to the return and made it fatal if one of those two fields isn't there too

pukkandan commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,55 @@
+
+
+class PrankCastIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?prankcast\.com/.*/showreel/(?P<id>\d+)-(?P<display_id>.+)'
```
```suggestion
    _VALID_URL = r'https?://(?:www\.)?prankcast\.com/[^/?#]+/showreel/(?P<id>\d+)-(?P<display_id>[^/?#]+)'
```

pukkandan commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,55 @@
+
+        webpage = self._download_webpage(url, video_id)
+
+        # Extract the JSON
```
Unnecessary comment. Same below

pukkandan commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,55 @@
+
+        # Get the co-hosts/guests
+        guests_json = self._parse_json(json_info.get('guests_json') or '{}', video_id)
+        guests = [] if not uploader else [uploader] + [x.get('name') for x in guests_json if x.get('name') is not None]
```
```suggestion
        guests = list(filter(None, [uploader] + traverse_obj(guests_json, (..., 'name'))))
```

pukkandan commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,55 @@
+
+        # Get dates
+        start_date = parse_iso8601(json_info.get('start_date'))
+        end_date = parse_iso8601(json_info.get('end_date'))
+
+        # Parse the duration of the stream
+        parsed_duration = None
+        if start_date is not None and end_date is not None:
+            parsed_duration = (end_date - start_date)
```
```suggestion
        parsed_duration = try_call(lambda: parse_iso8601(json_info['end_date']) - start_date)
```

pukkandan commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,55 @@
+            'timestamp': start_date,
+            'uploader': uploader,
+            'channel_id': json_info.get('user_id'),
+            'duration': parsed_duration,
+            'cast': guests,
```
These can be inlined

pukkandan commented on yt_dlp/extractor/prankcast.py:
```diff
@@ -0,0 +1,55 @@
+class PrankCastIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?prankcast\.com/.*/showreel/(?P<id>\d+)-(?P<display_id>.+)'
+    _TESTS = [{
+        'url': 'https://prankcast.com/Devonanustart/showreel/1561-Beverly-is-back-like-a-heart-attack-',
```
Test does not pass

columndeeply commented:
Done, I've added the missing fields to the test and changed all the other points.
